### ISSUE TITLE
Null dereference warning within local function

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#38548]
The compiler generates a warning for `y.ToString()` in 16.4 but no warning in 16.3.
```C#
class Program
{
    static void F(System.Action a) { }

    static void M(string? x)
    {
        string? y = x;
        if (y == null) return;

        F(Local);
        return;

        void Local()
        {
            y.ToString();
        }
    }
}
```

**Comment by jnm2** [comment]
Updating to 16.4 caused a new incorrect warning to break a build. The repro is the same as this issue.

**Comment by andrewsav-datacom** [comment]
Another repo:

```csharp
#nullable enable
using System;
using System.Collections.Generic;

namespace Nullability
{
    class Program
    {
        static void Main(string[] args)
        {
            var s = new List<int>();
            var b = s == null;
            var i = s.Count;
        }
    }
}
```

**Comment by jnm2** [comment]
@andrewsav-datacom I think that one is by design. In any case, it doesn't involve local functions.

**Comment by andrewsav-datacom** [comment]
@jnm2, right. Can you clarify about the "by design" part - otherwise it may be worthwhile for me opening another issue about that.

**Comment by jnm2** [comment]
Null checks are often a useful way of telling the compiler to reset its prior inference about a variable's nullability. In this example it's obviously not useful, but that's only because `s == null` itself is not a useful thing to write.

**Comment by andrewsav-datacom** [comment]
@jnm2 I'm sorry I do not want to derail the thread, so if you can suggest more appropriate avenue for this discussion, you will be welcome, but something is still not clear.

If variable s is not of a nullable type, what is the point of it being compared with null. I would expect error / warning on _that_ line instead.

When use say " a useful way of telling the compiler to reset its prior inference" it implies that the inference could be wrong. How can that happen?

**Comment by jnm2** [comment]
@andrewsav-datacom I'm not sure. It may be worth opening an issue to get a full answer. From my observations, inference doesn't always have enough info to go on. The variable's nullability may be inferred from something that is null-oblivious or is just as tenuous as the null check. It does seem like it would be nice for a warning or suggestion to be shown in your example where there is no uncertainty about the prior nullability.

**Comment by jcouv** [comment]
Warning is no longer produced in 16.5p2. I think this was fixed by https://github.com/dotnet/roslyn/pull/40422
I'll go ahead and close the issue. Thanks
